clang/lib/CIR/Dialect/Transforms/LoweringPrepare.cpp

```diff
@@ -1120,6 +1125,79 @@ void LoweringPreparePass::lowerIterEndOp(IterEndOp op) {
   op.erase();
 }
 
+/// lowering construction of std::initializer_list.
+/// 1. alloca array for arg list.
+/// 2. copy arg list to array.
+/// 3. construct std::initializer_list from array.
+void LoweringPreparePass::lowerStdInitializerListOp(StdInitializerListOp op) {
+  auto loc = op.getLoc();
+  cir::CIRDataLayout dataLayout(theModule);
+  auto args = op.getArgs();
+
+  auto stdInitializerListType = mlir::cast<cir::StructType>(
+      mlir::cast<cir::PointerType>(op.getInitList().getType()).getPointee());
+  clang::RecordDecl::field_range stdInitializerListFields =
+      stdInitializerListType.getAst().getRawDecl()->fields();
+
+  mlir::Type elementType =
+      mlir::cast<cir::PointerType>(stdInitializerListType.getMembers()[0])
+          .getPointee();
+  auto tempArrayType =
+      cir::ArrayType::get(&getContext(), elementType, args.size());
+
+  CIRBaseBuilderTy builder(getContext());
+  builder.setInsertionPointAfter(op);
+
+  IntegerAttr alignment = builder.getI64IntegerAttr(
+      dataLayout.getPrefTypeAlign(tempArrayType).value());
+  assert(!cir::MissingFeatures::addressSpace());
+  mlir::Value arrayPtr = builder.createAlloca(
+      loc, cir::PointerType::get(tempArrayType), tempArrayType, "", alignment);
+  mlir::Value arrayStartPtr =
+      builder.createCast(cir::CastKind::array_to_ptrdecay, arrayPtr,
+                         cir::PointerType::get(elementType));
+  for (unsigned i = 0; i < args.size(); i++) {
+    if (i == 0) {
+      builder.createStore(loc, args[i], arrayStartPtr);
+    } else {
+      mlir::Value offset = builder.getUnsignedInt(loc, i, 64);
+      mlir::Value dest = builder.create<cir::PtrStrideOp>(
+          loc, arrayStartPtr.getType(), arrayStartPtr, offset);
+      builder.createStore(loc, args[i], dest);
+    }
+  }
+
+  // FIXME(cir): better handling according to different field type. [ptr ptr],
+  // [ptr size], [size ptr].
```

Comment:
What's this FIXME referring to? You're handling both `[ptr ptr]` and `[ptr size]` below, and I don't think `[size ptr]` is used. What improvements did you have in mind?

Reply:
According to [standard](https://eel.is/c++draft/support.initlist#initializer), it is not define the member of `std::initializer_list`. So it can be [ptr ptr], [ptr size]. But it is also possible to put the size as first member and ptr as second member.